doc: more details in ZoneConcierge's documentation (#422)
babylonchain/babylon: x/zoneconcierge/README.md

# ZoneConcierge

> The Zone Concierge module is responsible for generating BTC timestamps of

Reviewer: ```suggestion
The Zone Concierge module is responsible for BTC time-stamping
```
?

Author: I feel the original version is better since "BTC timestamping" does not look like a verb, wdyt?

Reviewer: OK with me.

headers from other PoS blockchains. These BTC timestamps allow PoS blockchains
integrating with Babylon to achieve Bitcoin security, i.e., forking the PoS
blockchain is as hard as forking Bitcoin. The Zone Concierge module leverages

> the IBC light client protocol to receive PoS blockchains' headers.

Reviewer 2: ```suggestion
the IBC protocol to receive PoS blockchains' headers and
provide them with succint and provable information about their timestamps.
```

There are two phases of integration for a PoS blockchain:

- **Phase 1 integration:** Babylon receives PoS blockchain headers via standard
  `MsgUpdateClient` messages in IBC light client protocol, timestamps them, and
  functions as a canonical chain oracle for the PoS blockchain.
  [Babylonscan](https://babylonscan.io/) shows PoS blockchains with phase 1
  integration.

> - **Phase 2 integration:** In addition to phase 1, phase 2 allows a consumer
>   chain to receive BTC timestamps from Babylon via an IBC channel, such that the

Reviewer: ```suggestion
- **Phase 2 integration:** In addition to phase 1, phase 2 allows a PoS
 blockchain to receive BTC timestamps from Babylon via an IBC channel, such that the
```

  PoS blockchain can use BTC timestamps to detect and resolve forks, as well as
  other use cases such as Bitcoin-assisted fast unbonding.

## Table of contents

- [Table of contents](#table-of-contents)
- [Concepts](#concepts)
  - [Problem Statement](#problem-statement)
  - [Design](#design)
  - [Use cases](#use-cases)
- [State](#state)
  - [Parameters](#parameters)
  - [ChainInfo](#chaininfo)
  - [EpochChainInfo](#epochchaininfo)
  - [CanonicalChain](#canonicalchain)
  - [Fork](#fork)
  - [Params](#params)
- [PostHandler for intercepting IBC headers](#posthandler-for-intercepting-ibc-headers)
- [Hooks](#hooks)
  - [Indexing headers upon `AfterEpochEnds`](#indexing-headers-upon-afterepochends)
  - [Sending BTC timestamps upon `AfterRawCheckpointFinalized`](#sending-btc-timestamps-upon-afterrawcheckpointfinalized)
- [Interaction with PoS blockchains under phase 1 integration](#interaction-with-pos-blockchains-under-phase-1-integration)
- [Interaction with PoS blockchains under phase 2 integration](#interaction-with-pos-blockchains-under-phase-2-integration)
- [Queries](#queries)

## Concepts

The Zone Concierge module is responsible for providing BTC timestamps of headers

> from other PoS blockchains. These BTC timestamps allow PoS blockchains to

Reviewer 2: ```suggestion
from PoS blockchains connected to Babylon via the IBC protocol.
These BTC timestamps allow PoS blockchains to
```

achieve Bitcoin security, i.e., forking a PoS blockchain is as hard as forking
Bitcoin. The Zone Concierge module leverages the IBC light client protocol to
maintain headers with a valid quorum certificate from PoS blockchains. These
headers receive Bitcoin timestamps together with the Babylon blockchain, thereby

> achieving Bitcoin security.

Reviewer 2: Maybe also note that we propagate information about the timestamps to the chains

### Problem Statement

> Babylon aims at providing Bitcoin security to other PoS blockchains. This

Reviewer 2: ```suggestion
Babylon aims to provide Bitcoin security to other PoS blockchains. This
```

involves two functionalities: 1) checkpointing Babylon to Bitcoin, and 2)
checkpointing other PoS blockchains to Babylon. The {[Epoching](../epoching/),
[Checkpointing](../checkpointing/), [BTCCheckpoint](../btccheckpoint/),
[BTCLightclient](../btclightclient/)} modules jointly provide the functionality
of checkpointing Babylon to Bitcoin. The Zone Concierge module and the [IBC
modules](https://github.com/cosmos/ibc-go) jointly provide the functionality of
checkpointing PoS blockchains to Babylon.

In order to checkpoint PoS blockchains to Babylon, Babylon needs to receive

> headers of PoS blockchains, and maintain all headers that have a *quorum

Reviewer 2: ```suggestion
headers of PoS blockchains and maintain all headers that have a *quorum
```

majority attacks.

To summarize, the Zone Concierge module aims at providing the following
guarantees:

- **Timestamping headers:** Babylon checkpoints PoS blockchains' (canonical and
  fork) headers with a valid quorum certificate.
- **Verifiability of timestamps:** Babylon can provide a proof that a given
  header is checkpointed by Bitcoin, where the proof is publicly verifiable
  assuming access to a BTC light client.

under the following assumptions:

- BTC is always secure with the [k-deep confirmation
  rule](https://en.bitcoin.it/wiki/Confirmation);
- There exists >=1 honest IBC relayer and >=1 vigilante {submitter, reporter};
  and
- The network is synchronous (i.e., messages are delivered within a known and
  finite time bound).

Note that the Bitcoin timestamping protocol uses Bitcoin as a single source of
truth, and does not make any assumption on the fraction of adversarial
validators in Babylon or PoS blockchains. That is, the above statement shall
hold even if Babylon and a PoS blockchain have dishonest supermajority. The
formal security analysis of the Bitcoin timestamping protocol can be found at

> the [S\&P'23 paper](https://arxiv.org/pdf/2207.08392.pdf).

Author: S&P'23 refers to [this conference](https://sp2023.ieee-security.org/) where the Bitcoin timestamping paper linked here is published. Will replace it with the paper title

> the [S\&P'23 paper](https://arxiv.org/pdf/2207.08392.pdf).

Reviewer 2: Would keep the S&P text though, it's the most prestigious crypto conference :)

### Design

The Zone Concierge module is responsible for checkpointing headers from PoS

> blockchains. Specifically, the Zone Concierge module

Reviewer 2: ```suggestion
blockchains and propagating succinct and verifiable information about them back to the PoS chain. Specifically, the Zone Concierge module
```

- leverages IBC light clients for checkpointing PoS blockchains;

> - intercepts and indexes headers from PoS blockchains; and

Author: I would keep "and" since 1) it shows all bullet points need to be true, and 2) it's more like a sentence

> - provides proofs that a header is checkpointed by Babylon and Bitcoin.

Reviewer 2: Note that this can also happen through propagating info through IBC.

one honest validator in the header's quorum certificate, and this header is
agreed by all honest validators.

Babylon leverages the IBC light client protocol to checkpoint PoS blockchains to
itself. In particular, each header with a valid quorum certificate can be viewed
as a timestamp, and Babylon can generate an inclusion proof that a given header
of a PoS blockchain is committed to Babylon's `AppHash`.

**Intercepting and Indexing Headers from PoS blockchains.** In order to further
checkpoint headers of PoS blockchains to Bitcoin, the Zone Concierge module
builds an index recording headers' positions on Babylon's ledger, which will
eventually be checkpointed by Bitcoin. To this end, the Zone Concierge module
intercepts headers from IBC light clients via a
[PostHandler](https://docs.cosmos.network/v0.50/learn/advanced/baseapp#runtx-antehandler-runmsgs-posthandler),
and indexes them.

Note that the Zone Concierge module intercepts all headers that have a valid
quorum certificate, including canonical headers and fork headers. A fork header
with a valid quorum certificate is a signal of the dishonest majority attack:

> the majority of validators are honest and sign conflicted headers.

Reviewer: ```suggestion
the majority of validators are dishonest and sign conflicted headers.
```
?

**Providing Proofs that a Header is Checkpointed by Bitcoin.** To supports use
cases that need to verify BTC timestamps of headers, Zone Concierge can provide

> proofs that the headers are indeed checkpointed to Bitocin. The proof includes

Reviewer: ```suggestion
proofs that the headers are indeed checkpointed to Bitcoin. The proof includes
```

### EpochChainInfo

The [epoch chain info storage](./keeper/epoch_chain_info_indexer.go) maintains
`ChainInfo` at the end of each Babylon epoch for each PoS blockchain. The key is
the PoS blockchain's `ChainID` plus the epoch number, and the value is a
`ChainInfo` object.

### CanonicalChain

The [canonical chain storage](./keeper/canonical_chain_indexer.go) maintains the
metadata of canonical IBC headers of a PoS blockchain. The key is the consumer
chain's `ChainID` plus the height, and the value is a `IndexedHeader` object.

> `IndexedHeader` is a structure storing an IBC header's metadata.

Reviewer: ```suggestion
`IndexedHeader` is a structure storing IBC header's metadata.
```

  uint64 height = 3;
  // time is the timestamp of this header on CZ ledger
  // it is needed for CZ to unbond all mature validators/delegations

>   // before this timestamp when this header is BTC-finalized

Reviewer: Do we favor American spelling everywhere then?

Author: I don't think there is any preference. Personally I use British English everywhere, except for (public facing) code since American English is somehow more popular in the ecosystem. Your call

Reviewer: I for one prefer British to American English.

}
```

Upon `AfterRawCheckpointFinalized` is triggered, the Zone Concierge module will
send an IBC packet including a `BTCTimestamp` to each PoS blockchain doing phase

> 2/3 integration with Babylon. The logic is defined at

Reviewer 2: ```suggestion
2 integration with Babylon. The logic is defined at
```
We don't separate these phases now, no?

Author: No we don't separate them anymore. Phase3 is an entirely client thing (i.e., a consumer chain can decide how to use a BTC timestamp without Babylon's cooperation). Removed 3

> [x/zoneconcierge/keeper/hooks.go](./keeper/hooks.go) and works as follows.

Reviewer: All this section is pretty good, but AFAIK it only applies to Phase 2. It's a little confusing to find it before the Phase 2 section below. Perhaps a restructuring, in which this description is consolidated under Phase 2 below would make sense.

That, or referencing this section below, and also putting a comment here about phase 2 applicability.

Author: Added statements that this is only used in phase2 integrtaion. Given that this section is the spec of zone concierge module I prefer to keep the spec of hooks here. Will extend the phase 1/2 sections to make them more informative and coherent with the spec.

1. Find all open IBC channels with Babylon's Zone Concierge module. The
   counterparty at each IBC channel is a PoS blockchain.
2. Get all BTC headers to be sent in BTC timestamps. Specifically,
   1. Find the segment of BTC headers sent upon last time
      `AfterRawCheckpointFinalized` is triggered.
   2. If all BTC headers in the segment are no longer canonical, the BTC headers
      to be sent will be the last `w+1` ones in the BTC light client, where

>       where `w` is the `checkpoint_finalization_timeout`

Reviewer 2: ```suggestion
     `w` is the `checkpoint_finalization_timeout`
```

      [parameter](../../proto/babylon/btccheckpoint/v1/params.proto) in the
      [BTCCheckpoint](../btccheckpoint/) module.
   3. Otherwise, the BTC headers to be sent will be from the latest header that
      is still canonical in the segment to the current tip of the BTC light
      client.
3. For each of these IBC channels:
   1. Find the `ChainID` of the counterparty chain (i.e., the PoS blockchain) in
      the IBC channel.
   2. Get the `ChainInfo` of the `ChainID` at the last finalized epoch.
   3. Get the metadata of the last finalized epoch and its corresponding raw
      checkpoint.
   4. Generate the proof that the last PoS blockchain's canonical header is
      committed to the epoch's metadata.
   5. Generate the proof that the epoch is sealed, i.e., receives a BLS
      multisignature generated by validators with >2/3 total voting power at the
      last finalized epoch.
   6. Generate the proof that the epoch's checkpoint is submitted, i.e., encoded
      in transactions on Bitcoin.
   7. Assemble all the above and the BTC headers obtained in step 2 as
      `BTCTimestamp`, and send it to the IBC channel in an IBC packet.

> ## Interaction with PoS blockchains under phase 1 integration

Author: Agreed. I see two paths of the below two sections and am not sure which to go:

1. Extending the two sections with more technical details and more connections with the spec above.
2. Keeping them high-level and moving them to the concepts section.

Right now I prefer 1 and will do this in a subsequent PR. Wdyt?

<!-- TODO: mermaid flowchart for the interaction -->

In phase 1 integration, Babylon maintains headers for a PoS blockchain via the
IBC light client protocol. The IBC light client of the PoS blockchain is
checkpointed by Bitcoin via Babylon, thus achieves Bitcoin security.

Babylon utilizes [IBC light client
protocol](https://github.com/cosmos/ibc/tree/main/spec/client/ics-007-tendermint-client)
for receiving headers from other PoS blockchains. The IBC headers are
encapsulated in IBC protocol's `MsgUpdateClient`
[message](https://github.com/cosmos/ibc-go/blob/v8.0.0/proto/ibc/core/client/v1/tx.proto#L20-L21),

> and is sent to the [IBC client

Reviewer: ```suggestion
and are sent to the [IBC client
```

module](https://github.com/cosmos/ibc-go/tree/v8.0.0/modules/core/02-client) by
an [IBC
relayer](https://github.com/cosmos/ibc/blob/main/spec/relayer/ics-018-relayer-algorithms/README.md).
The `IBCHeaderDecorator` PostHandler intercepts the headers and index their
positions in `ChainInfo` storage, as per
[here](#indexing-headers-upon-afterepochends). This effectively checkpoints the

> headers of PoS blockchains, completing the phased 1 integration.

Reviewer: ```suggestion
headers of PoS blockchains, completing the Phase 1 integration.
```

timestamps of headers back to each PoS blockchain. Each PoS blockchain can
verify the BTC timestamp and ensure that each header is finalized by Bitcoin,
thus obtaining Bitcoin security. The BTC timestamps will allow a PoS blockchain
to make use of BTC timestamps for different use cases, such as BTC-assisted fast
unbonding.

The phase 2 integration does not require any change to the PoS blockchain's
code. Rather, it only needs to deploy a [Babylon
contract](https://github.com/babylonchain/babylon-contract) on the PoS
blockchain, and start an IBC relayer between Babylon and the Babylon contract on
the PoS blockchain. The Babylon contract can be deployed to a blockchain
supporting [CosmWasm](https://github.com/CosmWasm/cosmwasm) smart contracts,
connects with Babylon's Zone Concierge module via an IBC channel, and receives
BTC timestamps from Babylon to help the PoS blockchain get Bitcoin security.

Upon a Babylon epoch is finalized, i.e., upon `AfterRawCheckpointFinalized` is
triggered, Babylon will send an IBC packet including a `BTCTimestamp` to each
PoS blockchain doing phase 2/3 integration with Babylon, as per
[here](#sending-btc-timestamps-upon-afterrawcheckpointfinalized).

> Note that Zone Concierge provides 1-to-all connection, where where the Zone

Reviewer: ```suggestion
Note that Zone Concierge provides 1-to-all connection, where the Zone
```

Concierge module establishes an IBC channel with each of multiple consumer
chains. Zone Concierge will send an BTC timestamp to each of these consumer
chains upon an epoch is finalized.

> ## Queries
> 
> The Zone Concierge module only has one message `MsgUpdateParams` for updating
> the module parameters via a governance proposal. It provides a set of queries

Reviewer: ```suggestion
## Messages and Queries

The Zone Concierge module only has one message `MsgUpdateParams`, for updating
the module parameters via a governance proposal.

It also provides a set of queries
```
